### ISSUE TITLE
Unfortunate parser response to missing opening square bracket

### ISSUE THREAD

**Issue by bwilkerson** [dart-lang/sdk#41000]
This was originally discovered while a user was editing a Flutter build method. Their objective was to replace one class of widget that takes a list of children to a class of widget that takes a fixed number of widgets. The steps were to change the name of the widget class in the constructor, remove the `children:` label and surrounding list literal delimiters, and then add labels before the constructors that were previously in the list.

The following code illustrates the problem:
```
class C {
  C({c});
  C m() {
    return C(
      c: [
        C(c: [ // here
          C(),
          C(),
          C(),
        ]),
        C(),
        C(),
      ],
    );
  }
}

var c = C(
  c: [
    C(c: [ // here
      C(),
      C(),
      C(),
    ]),
    C(),
    C(),
  ],
);
```
At either of the two marked lines, comment out either `c: [` or `[`. Each of those four possibilities generate more diagnostics than might be expected from removing one or two tokens, and none of the generated diagnostics suggests a valid solution (either adding back a `[` or removing the unmatched ']'). Note that not all of the diagnostics are from the parser.

The problem is that the parser isn't able to detect that a problem exists until long past the point at which the problem actually occurs. I don't actually expect that it will be able to detect the problem earlier because I believe that would require backtracking, which this parser is not designed to support. However, it would be nice if it could, at some point, detect that the problem exists and stop generating follow-on errors from that point on.

For example, if it could determine that there's a problem when it encounters the unmatched closing bracket, then perhaps it could recover in a way that would make not generate any diagnostics after that point.

I haven't tried, but it's likely that there are similar problems with missing curly braces (`{`) and parentheses (`(`).

**Comment by bwilkerson** [comment]
@jensjoha @johnniwinther

**Comment by jensjoha** [comment]
Should be fixed with 1f9b5c199659ab9ab295457ffeadb783dcfbe3de.

**Comment by stereotype441** [comment]
Reopening since I had to revert the fix - see 275b922ee88502ee12d34af8b0476291ae6eb7b4

**Comment by stereotype441** [comment]
Fix was re-applied in 3ca37fb66e5fabcf52f75b25189749c943120156
